syntax = "proto2";

package circlechef;

//
// Initial version
//  - Our initial version
//
// Version 1
//  - Backward compatible with Initial version
//  - Added Graph to represent sub graphs
//  - Added name, version(default as 1), graph in ModelRecipe
//

// This enum value corresponds to TensorType in TensorFlow Lite schema
enum TensorType {
  FLOAT32 = 0;
  INT32 = 2;
  UINT8 = 3;
  INT64 = 4;
  BOOL = 6;
}

message TensorShape {
  repeated uint32 dim = 3;
}

message TensorFiller {
  optional string tag = 1;
  repeated string arg = 2;
}

message TensorQuantization {
  repeated float min = 1;
  repeated float max = 2;
  repeated float scale = 3;
  repeated int64 zero_point = 4;
}

message Operand {
  optional string name = 1;
  optional TensorType type = 2;
  optional TensorShape shape = 3;
  optional TensorFiller filler = 4;
  optional TensorQuantization quant = 5;
}

// This enum value corresponds to Padding in TensorFlow Lite schema
enum Padding {
  SAME = 0;
  VALID = 1;
}

// This enum value corresponds to ActivationFunctionType in TensorFlow Lite schema
enum Activation {
  NONE = 0;
  RELU = 1;
  RELU6 = 3;
}

message BatchMatMulOptions {
  optional bool adjoint_lhs = 1 [default = false];
  optional bool adjoint_rhs = 2 [default = false];
}

message Operation {
  optional string type = 1;
  repeated string input = 2;
  repeated string output = 3;

  optional BatchMatMulOptions batch_matmul_options = 100;
}

// For additional subgraphs
message Graph {
  repeated Operand operand = 1;
  repeated Operation operation = 2;
  repeated string input = 3;
  repeated string output = 4;
  optional string name = 5;
}

message ModelRecipe {
  repeated Operand operand = 1;
  repeated Operation operation = 2;
  repeated string input = 3;
  repeated string output = 4;
  optional string name = 5;
  optional uint32 version = 6 [default = 1];
  repeated Graph graph = 7;
}
